language/bytecode-verifier/src/borrow_graph.rs

```diff
@@ -0,0 +1,272 @@
+        }
+    }
+
+    /// removes `nonce` and appropriately concatenates each incoming edge with each outgoing edge of `nonce`
+    pub fn remove_nonce(&mut self, nonce: Nonce) {
+        checked_assume!(self.invariant());
+        let nonce_edge_set = self.0.remove(&nonce).unwrap();
+        let removed_edges = {
+            let mut x = BTreeMap::new();
+            for (n, es) in &self.0 {
+                x.insert(
+                    n.clone(),
+                    es.iter()
+                        .filter(|x| x.to == nonce)
+                        .cloned()
+                        .collect::<Vec<_>>(),
+                );
+            }
+            x
+        };
+        for (n, es) in &removed_edges {
+            es.iter().for_each(|removed_edge| {
+                let n_edge_set_ref = self.0.get_mut(n).unwrap();
+                n_edge_set_ref.remove(removed_edge);
+                nonce_edge_set.iter().for_each(|nonce_edge| {
+                    if removed_edge.edge_type == EdgeType::Strong {
+                        let mut new_label = vec![];
+                        new_label.append(&mut removed_edge.label.clone());
+                        new_label.append(&mut nonce_edge.label.clone());
+                        let edge = Edge {
+                            edge_type: nonce_edge.edge_type,
+                            label: new_label,
+                            to: nonce_edge.to,
+                        };
+                        n_edge_set_ref.insert(edge);
+                    } else {
+                        let edge = Edge {
+                            edge_type: EdgeType::Weak,
+                            label: removed_edge.label.clone(),
+                            to: nonce_edge.to,
+                        };
+                        n_edge_set_ref.insert(edge);
+                    }
+                });
+            });
+        }
+        checked_verify!(self.invariant());
+        checked_postcondition!(!self.0.contains_key(&nonce));
+    }
+
+    /// renames nonces in `self` according to `nonce_map`
+    pub fn rename_nonces(&self, nonce_map: BTreeMap<Nonce, Nonce>) -> Self {
+        checked_assume!(self.invariant());
+        let mut new_graph = BTreeMap::new();
+        for (n, es) in &self.0 {
+            new_graph.insert(
+                nonce_map[n],
+                es.iter()
+                    .map(|x| Edge {
+                        edge_type: x.edge_type,
+                        label: x.label.clone(),
+                        to: nonce_map[&x.to],
+                    })
+                    .collect::<BTreeSet<_>>(),
+            );
+        }
+        let new_borrow_graph = BorrowGraph(new_graph);
+        checked_verify!(self.invariant());
+        checked_verify!(new_borrow_graph.invariant());
+        new_borrow_graph
+    }
+
+    /// checks if `self` covers `other`
+    pub fn abstracts(&self, other: &Self) -> bool {
```

Comment:
Why did this become `abstracts`? Neither is more abstract than each other. One might be more restrictive than the other (i.e. a weakening of the other). But I don't feel like abstracts applies here

Reply:
This is "abstracts" in the sense of abstract interpretation.  We could use "weakens" also.  I asked @hermanventer what he preferred and did his bidding.

Reply:
I dropped "subsumes" because I thought it would not make sense to anybody.

Reply:
That's not what the abstract in abstract in interpretation refers to, which is why I find this confusing